src/github.com/getlantern/fronted/server.go

```diff
@@ -169,27 +174,63 @@ func (server *Server) Serve(l net.Listener) error {
 // in a countingConn if an InstanceId was configured.
 func (server *Server) dialDestination(addr string) (net.Conn, error) {
 	if !server.AllowNonGlobalDestinations {
-		host, _, err := net.SplitHostPort(addr)
+		err := server.checkForLocalAddress(addr)
 		if err != nil {
-			err = fmt.Errorf("Unable to split host and port for %v: %v", addr, err)
 			log.Error(err.Error())
 			return nil, err
 		}
-		ipAddr, err := net.ResolveIPAddr("ip", host)
+	}
+
+	if server.AllowedPorts != nil {
+		err := server.checkForDisallowedPort(addr)
 		if err != nil {
-			err = fmt.Errorf("Unable to resolve destination IP addr: %s", err)
-			log.Error(err.Error())
-			return nil, err
-		}
-		if !ipAddr.IP.IsGlobalUnicast() {
-			err = fmt.Errorf("Not accepting connections to non-global address: %s", host)
 			log.Error(err.Error())
 			return nil, err
 		}
 	}
+
 	return net.DialTimeout("tcp", addr, dialTimeout)
 }
 
+func (server *Server) checkForLocalAddress(addr string) error {
+	host, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		return fmt.Errorf("Unable to split host and port for %v: %v", addr, err)
+	}
+
+	ipAddr, err := net.ResolveIPAddr("ip", host)
+	if err != nil || len(ipAddr.IP) == 0 {
+		return fmt.Errorf("Unable to resolve destination IP addr %v: %v", host, err)
+	}
+
+	if !ipAddr.IP.IsGlobalUnicast() {
+		return fmt.Errorf("Not accepting connections to non-global address: %s", addr)
+	}
+
+	return nil
+}
+
+func (server *Server) checkForDisallowedPort(addr string) error {
+	_, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return fmt.Errorf("Unable to split host and port for %v: %v", addr, err)
+	}
+
+	portAllowed := false
+	for _, allowed := range server.AllowedPorts {
+		if port == strconv.Itoa(allowed) {
```

Comment:
Maybe slightly better to do this outside of the loop, as in -- not a biggie of course, but since every request hits this:

```go
port := strconv.Atoi(portString)
for _, allowed := range server.AllowedPorts {
		if port == allowed {
```